### PR TITLE
Decouple semantic segmentation models

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

Eliminate mmseg dependency for models, decouple all models, introduce updated model structure, added support for torchvision data pipeline

Unit tests will be added with next PRs.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by kprokofi** [comment]
Strange error on CI with CUDA
image

Locally, all unit tests are passed with tox
image

File: src/otx/algo/segmentation/heads/custom_fcn_head.py
Comment (jaegukhyun) [inline src/otx/algo/segmentation/heads/custom_fcn_head.py:1]: I think now we can change name to fcn_head.py. What do you think?

File: src/otx/algo/segmentation/heads/custom_ham_head.py
Comment (jaegukhyun) [inline src/otx/algo/segmentation/heads/custom_ham_head.py:1]: Also this file could change name

File: src/otx/algo/segmentation/losses/cross_entropy_loss_with_ignore.py
Comment (jaegukhyun) [inline src/otx/algo/segmentation/losses/cross_entropy_loss_with_ignore.py:1]: Just for a note, many of modules in this file are duplicated with detection I think. Let's merge this after decoupling all tasks.

**Review by jaegukhyun — COMMENTED** [review]
Overall LGTM.
Could you share some regression tests between decoupling?
In some case torchvision data pipeline slower than mm's pipeline since mm use opencv and it is faster than torch tensor ops.
So we need regression tests results for accuracy and e2e time.

**Comment by jaegukhyun** [comment]
And it seems unit tests run on cpu machine. So I think you need to fix unit tests failure on cpu machine.

File: src/otx/algo/modules/conv_module.py [hunk 1]
```diff
@@ -355,3 +355,102 @@ def create_from_conv_bn(
         self.turn_on_efficient_conv_bn_eval(efficient_conv_bn_eval)
 
         return self
+
+
+class DepthwiseSeparableConvModule(nn.Module):
```
Comment (sungchul1) [inline src/otx/algo/modules/conv_module.py]: It is already implemented in develop. 
https://github.com/openvinotoolkit/training_extensions/blob/3d6bbdf673c1d1072dc9813c48cdcc3a241c976d/src/otx/algo/modules/depthwise_separable_conv_module.py#L14
Please remove it or move the previously implemented one to here.

File: src/otx/algo/segmentation/base_model.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import torch.nn.functional as f
+from torch import Tensor, nn
+
+from otx.algo.segmentation.losses import create_criterion
+
+if TYPE_CHECKING:
+    from otx.core.data.entity.base import ImageInfo
+
+
+class BaseSegmNNModel(nn.Module):
```
Comment (sungchul1) [inline src/otx/algo/segmentation/base_model.py]: What is `NN`? Is it from `nn.Module`? If so, why isn't it in `BaseSegmHead`?

File: src/otx/algo/segmentation/heads/custom_fcn_head.py [hunk 3]
```diff
@@ -5,98 +5,130 @@
...
+        return self.cls_seg(output)
 
 
-@MODELS.register_module()
-class CustomFCNHead(ClassIncrementalMixin, FCNHead):
+class CustomFCNHead(FCNHead):
     """Custom FCNHead implementation.
 
-    Custom FCNHead supports ignored label for class incremental learning cases.
+    Custom FCNHead supports iterative agregator and ignored label for class
```
Comment (sungchul1) [inline src/otx/algo/segmentation/heads/custom_fcn_head.py]:
```suggestion
    Custom FCNHead supports iterative aggregator and ignored label for class
```

File: src/otx/algo/segmentation/heads/custom_fcn_head.py [hunk 4]
```diff
@@ -5,98 +5,130 @@
...
-        return valid_label_mask
+    def forward(self, inputs: Tensor) -> Tensor:
+        """Forward function."""
+        output = self._forward_feature(inputs)
+        return self.cls_seg(output)
 
 
-@MODELS.register_module()
-class CustomFCNHead(ClassIncrementalMixin, FCNHead):
+class CustomFCNHead(FCNHead):
```
Comment (sungchul1) [inline src/otx/algo/segmentation/heads/custom_fcn_head.py]: Now, it seems to need to change this name to reveal more of its characteristics.

File: src/otx/algo/segmentation/litehrnet.py [hunk 5]
```diff
@@ -5,66 +5,530 @@
...
-        super().__init__(
-            label_info=label_info,
-            config=config,
-            optimizer=optimizer,
-            scheduler=scheduler,
-            metric=metric,
-            torch_compile=torch_compile,
-        )
+    @property
+    def ignore_scope(self) -> dict[str, str | dict[str, list[str]]]:
```
Comment (sungchul1) [inline src/otx/algo/segmentation/litehrnet.py:73]: Isn't there any way not to use these export parameters in the source code?
(I don't know the history of why it came here. If it was already agreed with others, feel free to ignore it.)

File: src/otx/algo/segmentation/modules/iterator.py
Comment (sungchul1) [inline src/otx/algo/segmentation/modules/iterator.py:1]: `aggregators.py` looks better than `iterator.py`.

File: src/otx/recipe/_base_/data/mmseg_base.yaml [hunk 6]
```diff
@@ -11,68 +11,70 @@ config:
...
-      - type: PhotoMetricDistortion
-      - type: Pad
-        size:
-          - 512
-          - 512
-        pad_val:
-          img: 0
-          seg: 255
-      - type: PackSegInputs
+      - class_path: torchvision.transforms.v2.RandomResizedCrop
```
Comment (sungchul1) [inline src/otx/recipe/_base_/data/mmseg_base.yaml:16]: It is a known issue that using torchvision operations shows slow performance.
What about replacing them with mmdet transforms implemented in torchvision pipeline like detection?

File: src/otx/algo/segmentation/segnext.py [hunk 7]
```diff
@@ -4,64 +4,124 @@
...
 
 if TYPE_CHECKING:
-    from lightning.pytorch.cli import LRSchedulerCallable, OptimizerCallable
+    from torch import nn
+
+
+class SegNextB(BaseSegmNNModel):
+    """SegNextB Model."""
+
+    default_backbone_configuration: ClassVar[dict[str, Any]] = {
```
Comment (sungchul1) [inline src/otx/algo/segmentation/segnext.py:23]: Is it required? They are already in the recipes.

Comment (kprokofi) [reply src/otx/algo/segmentation/segnext.py:23]: It is more for API usage. If we don't add these defaults - user should pass this entirely to initialize the class. With this addition, we can just create model passing only base_model_name. Moreover, we can pass only several parameters via dict (for backbone or head) and it will override defaults. I thought it can be useful.

File: src/otx/recipe/_base_/data/mmseg_base.yaml (hunk 6 above)
Comment (kprokofi) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: Do we have it implemented in OTX? I will take a look

Comment (harimkang) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: https://github.com/openvinotoolkit/training_extensions/blob/3d6bbdf673c1d1072dc9813c48cdcc3a241c976d/src/otx/core/data/transform_libs/torchvision.py#L549C7-L549C24
However, this approach still has the issue of being slow. See also https://github.com/openvinotoolkit/training_extensions/issues/3379

File: src/otx/algo/segmentation/litehrnet.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/litehrnet.py:73]: Not sure what you mean, but ignore scope in such representation is required for LiteHRNet unfortunately, we can also add entire backbone to ignored scope, but it doesn't make sense to compress this model. It is from NNCF team

File: src/otx/algo/segmentation/heads/custom_fcn_head.py (hunk 4 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/heads/custom_fcn_head.py]: OK

File: src/otx/algo/segmentation/base_model.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/base_model.py]: I will remove NN

File: src/otx/recipe/_base_/data/mmseg_base.yaml (hunk 6 above)
Comment (sungchul1) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: Yeah, conversion parts in the pipeline will be fixed or the whole pipeline can be updated by using other transform library.
But for now, please refer previously implemented ones.

File: src/otx/algo/segmentation/segnext.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/algo/segmentation/segnext.py:23]: I agree with your intention. I have considered how to use it more cleanly, but I have no idea :)

File: src/otx/algo/segmentation/modules/iterator.py
Comment (kprokofi) [reply src/otx/algo/segmentation/modules/iterator.py:1]: done https://github.com/openvinotoolkit/training_extensions/pull/3376/commits/04c51228f2b7e7f32812e63920e6ee011a199fad

File: src/otx/algo/segmentation/heads/custom_fcn_head.py (hunk 4 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/heads/custom_fcn_head.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3376/commits/04c51228f2b7e7f32812e63920e6ee011a199fad

File: src/otx/algo/segmentation/heads/custom_fcn_head.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/heads/custom_fcn_head.py]: removed https://github.com/openvinotoolkit/training_extensions/pull/3376/commits/04c51228f2b7e7f32812e63920e6ee011a199fad

File: src/otx/algo/segmentation/base_model.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/base_model.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3376/commits/04c51228f2b7e7f32812e63920e6ee011a199fad

File: src/otx/algo/modules/conv_module.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/algo/modules/conv_module.py]: deleted https://github.com/openvinotoolkit/training_extensions/pull/3376/commits/04c51228f2b7e7f32812e63920e6ee011a199fad

File: src/otx/algo/segmentation/backbones/mscan.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+        Args:
+            drop_prob (float): Probability of the path to be zeroed. Default: 0.1
+        """
+        super().__init__()
+        self.drop_prob = drop_prob
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward function."""
+        return drop_path(x, self.drop_prob, self.training)
```
Comment (eugene123tw) [inline src/otx/algo/segmentation/backbones/mscan.py:53]: Should we keep DropPath? It appears redundant alongside timm models (an existing dependency). Can we substitute it with timm.models.layers.DropPath for consistency?

Comment (kprokofi) [reply src/otx/algo/segmentation/backbones/mscan.py:53]: We should revisit everything what can be deleted / omitted. But maybe it is better to do after FF and basic functionality. Since we have a tight schedule...

File: src/otx/recipe/_base_/data/mmseg_base.yaml (hunk 6 above)
Comment (kprokofi) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: We don't support masks here unfortunately

Comment (kprokofi) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: I added support for masks resizing, but I don't see any difference in speed (train/data_time and iter_time is the same during training), but I see drop in accuracy and worse convergence. Let's decide on our augmentations backend in next PRs. For now, I would like to leave torchvisionv2 for segmentation.

**Comment by kprokofi** [comment]
Tests are passed. I propose to merge this PR, since today is FF. 
Please, feel free to leave comments to address in the following PRs before code freeze.

I compared algorithms on KITTI dataset.

iter_time is  the same, data_train_time is worse on 0.002 sec.,  due to difference in resizing and augmentations, we have changed accuracy values about 2-4 % DICE score on this dataset. We need to align augmentations with mm*

**Review by harimkang — APPROVED** [review]

File: src/otx/recipe/_base_/data/mmseg_base.yaml (hunk 6 above)
Comment (sungchul1) [reply src/otx/recipe/_base_/data/mmseg_base.yaml:16]: It is fine if there is little difference between mmdet and torcivision on semantic segmentation.
Let's discuss the future plan later as you said.

File: src/otx/algo/segmentation/litehrnet.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algo/segmentation/litehrnet.py:73]: I know why the `ignore_scope` is needed for export, but I'm not sure if they must be in here.
I think it would be better to separate it from the model and save it separately in a location related to the recipe (`openvino_model.yaml`?) like 1.x.
But I think my suggestion is not the scope for this PR, please consider it later.

**Review by sungchul1 — APPROVED** [review]
